src/openms/source/ANALYSIS/ID/FIAMSDataProcessor.cpp

```diff
@@ -0,0 +1,422 @@
+  if (this == &rhs) return *this;
+  filename_ = rhs.filename_;
+  dir_input_ = rhs.dir_input_;
+  dir_output_ = rhs.dir_output_;
+  resolution_ = rhs.resolution_;
+  polarity_ = rhs.polarity_;
+  min_mz_ = rhs.min_mz_;
+  max_mz_ = rhs.max_mz_;
+  bin_step_ = rhs.bin_step_;
+  db_mapping_ = rhs.db_mapping_;
+  db_struct_ = rhs.db_struct_;
+  positive_adducts_ = rhs.positive_adducts_;
+  negative_adducts_ = rhs.negative_adducts_;
+  store_progress_ = rhs.store_progress_;
+  mzs_ = rhs.mzs_;
+  bin_sizes_ = rhs.bin_sizes_;
+  return *this;
+}
+
+void FIAMSDataProcessor::cutForTime(const MSExperiment & experiment, vector<MSSpectrum> & output, float n_seconds) {
+    for (auto s : experiment.getSpectra()) {
+        if (s.getRT() < n_seconds) output.push_back(s);
+    }
+}
+
+MSSpectrum FIAMSDataProcessor::mergeAlongTime(
+  const std::vector<MSSpectrum> & input
+  ) {
+    MSSpectrum output;
+    for (size_t i = 1; i < mzs_.size() - 1; i++) {
+        OpenMS::MSSpectrum full_spectrum = OpenMS::SpectrumAddition::addUpSpectra(
+            input, bin_sizes_[i], false
+        );
+        for (auto it = full_spectrum.begin(); it != full_spectrum.end(); ++it) {
+            if (it->getMZ() > mzs_[i+1]) break;
+            if (it->getMZ() >= mzs_[i]) output.push_back(*it);
+        }
+    }
+    output.sortByPosition();
+    return output;
+}
+
+MSSpectrum FIAMSDataProcessor::extractPeaks(const MSSpectrum & input) {
+  MSSpectrum spectrum(input);
+
+  SavitzkyGolayFilter filter;
+  filter.filter(spectrum);
+
+  MorphologicalFilter morph_filter;
+  morph_filter.filter(spectrum);
+
+  PeakPickerHiRes picker;
+
+  MSSpectrum picked;
+  picker.pick(spectrum, picked);
+
+  return picked;
+}
+
+FeatureMap FIAMSDataProcessor::convertToFeatureMap(const MSSpectrum & input) {
+  FeatureMap output;
+  for (auto it = input.begin(); it != input.end(); ++it) {
+      Feature f;
+      f.setIntensity(it->getIntensity());
+      f.setMZ(it->getMZ());
+      output.push_back(f);
+      output[0].setMetaValue("scan_polarity", polarity_);
+  }
+  return output;
+}
+
+void FIAMSDataProcessor::runAccurateMassSearch(FeatureMap & input, OpenMS::MzTab & output) {
+  Param ams_param;
+  ams_param.setValue("ionization_mode", "auto");
+  ams_param.setValue("mass_error_value", 1e+06 / (resolution_*2));
+  ams_param.setValue("db:mapping", ListUtils::create<String>(db_mapping_));
+  ams_param.setValue("db:struct", ListUtils::create<String>(db_struct_));
+  ams_param.setValue("positive_adducts", positive_adducts_);
+  ams_param.setValue("negative_adducts", negative_adducts_);
+
+  AccurateMassSearchEngine ams;
+  ams.setParameters(ams_param);
+  ams.init();
+
+  ams.run(input, output);
+}
+
+MSSpectrum FIAMSDataProcessor::trackNoise(const MSSpectrum & input) {
+  SignalToNoiseEstimatorMedianRapid sne(50);
+  MSSpectrum output;
+  if (input.size() == 0) {
+    return output;
+  }
+  std::vector<double> mzs, intensities;
```

Comment:
I would recommend using `.reserve(...)` with std::vector for large vector sizes particularly if you know the size ahead of time.  MSSpectrum and other OpenMS types also use std::vector under the hood so there maybe opportunities in to use it there as well.